lib/cc/workspace/path_tree.rb

```diff
@@ -0,0 +1,115 @@
+require "pathname"
+require "set"
+
+module CC
+  class Workspace
+    class PathTree
+      def initialize(root_path, children = {})
+        @root_path = root_path.dup.freeze
+        @children = children
+      end
+
+      def clone
+        self.class.new(root_path, children.dup)
+      end
+
+      def exclude_paths(paths)
+        paths.each { |path| delete(normalized_path_pieces(path)) }
+      end
+
+      def include_paths(paths)
+        paths.each { |path| insert(normalized_path_pieces(path)) }
+      end
+
+      def all_paths
+        if populated?
+          children.values.flat_map(&:all_paths)
+        else
+          [File.join(root_path, File::SEPARATOR)]
+        end
+      end
+
+      protected
+
+      def populated?
+        children.present?
+      end
+
+      def delete(path_pieces)
+        populate_all
+
+        if 1 == path_pieces.size
+          children.delete(path_pieces[0])
+        else
+          child = children[path_pieces[0]]
+          if child
+            child.delete(path_pieces.drop(1))
+            children.delete(path_pieces[0]) unless child.populated?
+          else
+            CLI.debug(
+              "Couldn't exclude because part of path doesn't exist or was already excluded.",
+              exclude_path: File.join(path_pieces),
+              nonexistent_path: File.join(root_path, path_pieces[0]),
+            )
+          end
+        end
+      end
+
+      def insert(path_pieces)
+        return if path_pieces.empty?
+
+        entry = Pathname.new(root_path).children.detect { |c| c.basename.to_s == path_pieces[0] }
+        if entry && entry.directory?
+          children[entry.basename.to_s] = PathTree.new(entry.to_s)
```

Comment:
`PathTree.new` here and `self.class.new` at line 81. Probably should be consistent, I think I prefer `self.class.new` since it's robust against renames.